korge/src/commonMain/kotlin/korlibs/korge/component/StageComponent.kt

```diff
@@ -36,26 +36,28 @@ fun <T : View> T.onAttachDetach(onAttach: Views.(T) -> Unit = {}, onDetach: View
     return this
 }
 
+private val viewsToTrack = linkedSetOf<View>()
+
 /**
  * Enables the use of [StageComponent] components.
  */
-fun Views.registerStageComponent() {
+fun Views.registerStageComponent(view: View) {
+    viewsToTrack += view
+
     val EXTRA_ID = "Views.registerStageComponent"
     if (views.getExtra(EXTRA_ID) == true) return
     views.setExtra(EXTRA_ID, true)
+
     val componentsInStagePrev = FastArrayList<ViewStageComponent>()
     val componentsInStageCur = linkedSetOf<ViewStageComponent>()
     val componentsInStage = linkedSetOf<ViewStageComponent>()
-    val tempViews: FastArrayList<View> = FastArrayList()
     onBeforeRender {
         componentsInStagePrev.clear()
         componentsInStagePrev += componentsInStageCur
         componentsInStageCur.clear()
 
-        val stagedViews = getAllDescendantViews(stage, tempViews)
-
-        stagedViews.fastForEach { view ->
-            if (view.hasExtra(__VIEW_STAGE_COMPONENT_NAME)) {
+        viewsToTrack.forEach { view ->
+            if (view.hasExtra(__VIEW_STAGE_COMPONENT_NAME) && view.hasAncestor(stage)) {
```

Comment:
`view.hasAncestor(stage)` to check the view is still on stage.
I don't like this part 🤔 Any ideas?

Reply:
@BioRyajenka https://github.com/korlibs/korge/blob/90bc4c320eca3c5598fb58f7519366ea6a438fd1/korge/src/commonMain/kotlin/korlibs/korge/view/View.kt#L524-L525

Reply:
The `stage` reference is cached in each view, so that property won't iterate upwards.
It was an optimization in KorGE 4.0.0-alpha-1: https://github.com/korlibs/korge/commit/cdc63b4cdbac5a8a688595337d6f6427b83e0888

Reply:
What I check here, is that view is assigned to the stage. When its assigned then `view.hasAncestor(stage)` is true.

Yes stage is cached but I don't understand the other part: `property won't iterate upwards`. What do you mean by that? 

Thanks.

Reply:
Ohhh I see. I didn't see you propagate cache value downside as well. I will change the implementation then